pkg/asset/quota/openstack/openstack.go

```diff
@@ -11,44 +14,90 @@ import (
 // These constraints can be used to check if there is enough quota for creating a cluster
 // for the install config.
 func Constraints(ci *validation.CloudInfo, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
-	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
-	for i, m := range controlPlanes {
-		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	var constraints []quota.Constraint
+
+	for i := 0; i < len(controlPlanes); i++ {
+		constraints = append(constraints, machineConstraints(ci, &controlPlanes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(controlPlanes))))
+
+	for i := 0; i < len(computes); i++ {
+		constraints = append(constraints, machineSetConstraints(ci, &computes[i])...)
```

Comment:
Please help me understand.

Let's imagine the case of 3 MachineSets with one replica each, and each replica has a flavor that needs 16GB.

I am figuring the resulting `[]quota.Constraint`:
```Go
[]quota.Constraint{
  {RAM, 16},
  {RAM, 16},
  {RAM, 16},
}
```

So the checker will check three times that we have 16 GB total. Instead, I expected  `[]quota.Constraint` to look like this:

```Go
[]quota.Constraint{
  {RAM, 16*3},
}
```

What am I missing?

Reply:
You're missing the call to aggregate(constraints) on line 29 below. The aggregate call was already there, and I think it simplifies the code a bit. Basically it adds up all constraints against the same resource.